File: core/src/main/java/xtdb/vector/extensions/AbsentType.java
package xtdb.vector.extensions;

import org.apache.arrow.memory.BufferAllocator;
import org.apache.arrow.vector.FieldVector;
import org.apache.arrow.vector.types.pojo.ArrowType;
import org.apache.arrow.vector.types.pojo.ExtensionTypeRegistry;
import org.apache.arrow.vector.types.pojo.FieldType;

public class AbsentType extends XtExtensionType {
    public static final AbsentType INSTANCE = new AbsentType();

    static {
        ExtensionTypeRegistry.register(INSTANCE);
    }

    public AbsentType() {
        super("absent", Null.INSTANCE);
    }

    @Override
    protected ArrowType deserialize(String serializedData) {
        return INSTANCE;
    }

    @Override
    public FieldVector getNewVector(String name, FieldType fieldType, BufferAllocator allocator) {
        return new AbsentVector(name, allocator, fieldType);
    }
}

File: core/src/main/kotlin/xtdb/vector/extensions/AbsentType.kt
package xtdb.vector.extensions

import org.apache.arrow.memory.BufferAllocator
import org.apache.arrow.vector.types.pojo.ExtensionTypeRegistry
import org.apache.arrow.vector.types.pojo.FieldType

object AbsentType : XtExtensionType("absent", Null.INSTANCE) {
    init {
        ExtensionTypeRegistry.register(this)
    }

    override fun deserialize(serializedData: String) = this

    override fun getNewVector(name: String, fieldType: FieldType, allocator: BufferAllocator) =
        AbsentVector(name, allocator, fieldType)
}
